File: Advent2022/Day7/Day7.py
from typing import Optional, Any
from dataclasses import dataclass
from itertools import combinations

import aoc_utils as au


@dataclass
class ChangeDirCmd:
    target: str


@dataclass
class ListDir:
    pass


@dataclass
class DirEntry:
    name: str


@dataclass
class FileEntry:
    size: int
    name: str


def make_path(cwd: list[str], name: str) -> str:
    return "/".join(cwd) + "/" + name


def get_dirs(
    shell_stuff: list[ChangeDirCmd | ListDir | DirEntry | FileEntry],
) -> dict[str, int]:

    dirs = {"/": 0}
    cwd = ["/"]

    for shell_item in shell_stuff:
        match shell_item:
            case ChangeDirCmd() as cd_cmd:
                match cd_cmd.target:
                    case "/":
                        cwd = ["/"]
                    case "..":
                        if len(cwd) > 1:
                            cwd.pop()
                        else:
                            raise Exception("Tried to go above root")
                    case _:
                        cwd.append(make_path(cwd, cd_cmd.target))
            case ListDir():
                pass
            case DirEntry() as dir_entry:
                if (fullpath := make_path(cwd, dir_entry.name)) not in dirs:
                    dirs[fullpath] = 0
            case FileEntry() as file_entry:
                for dir in cwd:
                    dirs[dir] += file_entry.size
            case _:
                raise (f"Unknown object {shell_item} seen")
    return dirs


def find_space(shell_stuff: list[ChangeDirCmd | ListDir | DirEntry | FileEntry]) -> int:

    dirs = get_dirs(shell_stuff)
    return sum([dir for dir in dirs.values() if dir <= 100000])


def find_space2(
    shell_stuff: list[ChangeDirCmd | ListDir | DirEntry | FileEntry],
) -> int:

    dirs = get_dirs(shell_stuff)

    needed = 30000000 - (70000000 - dirs["/"])

    return min([dir for dir in dirs.values() if dir >= needed])


def test_one(
    shell_stuff: list[ChangeDirCmd | ListDir | DirEntry | FileEntry],
) -> Optional[int]:
    return find_space(shell_stuff)


# expected 1348005
def part_one(
    shell_stuff: list[ChangeDirCmd | ListDir | DirEntry | FileEntry],
) -> Optional[int]:
    return find_space(shell_stuff)


def test_two(
    shell_stuff: list[ChangeDirCmd | ListDir | DirEntry | FileEntry],
) -> Optional[int]:
    return find_space2(shell_stuff)


# expected 12785886
def part_two(
    shell_stuff: list[ChangeDirCmd | ListDir | DirEntry | FileEntry],
) -> Optional[int]:
    return find_space2(shell_stuff)


if __name__ == "__main__":
    shell_parser = au.RegexParser(
        [
            (r"\$ cd (.*)", lambda m: ChangeDirCmd(m[0])),
            (r"\$ ls", lambda m: ListDir()),
            (r"dir (.*)", lambda m: DirEntry(m[0])),
            (r"(\d*) (.*)", lambda m: FileEntry(int(m[0]), m[1])),
        ]
    )
    day = au.Day(
        2022,
        7,
        test_one,
        test_two,
        part_one,
        part_two,
        test_input=shell_parser,
        input=shell_parser,
    )

    day.run_all(run_tests=True)

File: Advent2022/Day7/Day7.cpp
#include "aoc_utils.h"

#include <string>
#include <regex>
#include <vector>
#include <map>
#include <deque>
#include <numeric>

using namespace std;

struct Command
{
    typedef enum
    {
        CD,
        LS,
        DIR,
        FILE
    } CMD_TYPE;

    CMD_TYPE m_type;
    string m_val1;
    string m_val2;

    Command(CMD_TYPE type, const string &val1 = "", const string &val2 = "") : m_type(type),
                                                                               m_val1(val1),
                                                                               m_val2(val2) {}
};

struct Input
{
    vector<Command> data;

    void push_back(const string &str)
    {
        static const regex changedir_re = regex("\\$ cd (.*)");
        static const regex listdir_re = regex("\\$ ls");
        static const regex direntry_re = regex("dir (.*)");
        static const regex fileentry_re = regex("(\\d*) (.*)");

        smatch m;
        if (regex_match(str, m, changedir_re))
            data.push_back(Command(Command::CD, m[1].str()));
        else if (regex_match(str, m, listdir_re))
            data.push_back(Command(Command::LS));
        else if (regex_match(str, m, direntry_re))
            data.push_back(Command(Command::DIR, m[1].str()));
        else if (regex_match(str, m, fileentry_re))
            data.push_back(Command(Command::FILE, m[1].str(), m[2].str()));
    }
};

class Day7 : public au::Day<Input, int>
{
private:
    string make_path(const vector<string> &cwd, const string &name)
    {
        if (cwd.back() == "/")
            return cwd.back() + name;
        else
            return cwd.back() + "/" + name;
    }

    map<string, int> get_dirs(const vector<Command> &shell_stuff)
    {
        map<string, int> dirs;
        vector<string> cwd{"/"};

        for (auto shell_item : shell_stuff)
        {
            switch (shell_item.m_type)
            {
            case Command::CD:
                if (shell_item.m_val1 == "/")
                    cwd = vector<string>{"/"};
                else if (shell_item.m_val1 == "..")
                {
                    if (cwd.size() > 1)
                        cwd.pop_back();
                }
                else
                    cwd.push_back(make_path(cwd, shell_item.m_val1));
                break;
            case Command::LS:
                break;
            case Command::DIR:
            {
                auto fullpath = make_path(cwd, shell_item.m_val1);
                if (dirs.find(fullpath) == dirs.end())
                    dirs[fullpath] = 0;
            }
            break;
            case Command::FILE:
                for (auto dir : cwd)
                    dirs[dir] += stoi(shell_item.m_val1);
                break;
            }
        }

        return dirs;
    }
    int find_space(const vector<Command> &shell_stuff)
    {
        auto dirs = get_dirs(shell_stuff);
        vector<int> sizes;

        transform(dirs.begin(),
                  dirs.end(),
                  back_inserter(sizes),
                  [](auto x)
                  { return x.second; });

        sizes.erase(remove_if(sizes.begin(), sizes.end(),
                              [](int x)
                              { return x > 100000; }),
                    sizes.end());

        return accumulate(sizes.begin(), sizes.end(), 0);
    }

    int find_space2(const vector<Command> &shell_stuff)
    {
        auto dirs = get_dirs(shell_stuff);

        int needed = 30000000 - (70000000 - dirs["/"]);

        vector<int> sizes;

        transform(dirs.begin(),
                  dirs.end(),
                  back_inserter(sizes),
                  [](auto x)
                  { return x.second; });

        sizes.erase(remove_if(sizes.begin(), sizes.end(),
                              [needed](int x)
                              { return x < needed; }),
                    sizes.end());
        return *min_element(sizes.begin(), sizes.end());
    }

public:
    Day7() : Day(2022, 7) {}

    int test_one(const Input &input)
    {
        return find_space(input.data);
    }

    int part_one(const Input &input)
    {
        return find_space(input.data);
    }

    int test_two(const Input &input)
    {
        return find_space2(input.data);
    }

    int part_two(const Input &input)
    {
        return find_space2(input.data);
    }
};

int main()
{
    Day7().run_all(true);

    return 0;
}
